### PR TITLE
Reading and Tree Op Refactor

### PR DESCRIPTION
**Pull request by charleslparker**
While integrating with wintermute, some performance things came up.  Specifically, the custom tree operator is apparently *really* expensive for tensorflow to serialize, so when there are dozens of them, serializing the network was taking half an hour.  I've basically folded multiple trees into the tree operator, so now it's basically an "ensemble operator".  This reduced serialization from 1800 seconds to 10 seconds.

(As an aside, this isn't the first tensorflow performance problem I've had.  It's mostly a nice thing, but lurking everywhere are scary inefficiencies that you can uncover by accident if your computation graph gets too big.  When you generate them programmatically, this can easily happen by accident).

I've also refactored reading here so it's more a part of the client's responsibility; the image reading code in tensorflow was just not extensible or flexible enough to handle most use cases.  The model wrappers (or client code for other bindings) will now take care of reading the image so tensorflow itself only deals with arrays of pixels.

### CODE REVIEW

**Comment by charleslparker** [comment]
@unmonoqueteclea - Unfortunately, this meant I had to add a dependency on PIL to read the images.  It's common enough though that I don't think it will cause major problems.

File: sensenet/load.py [hunk 1]
```diff
@@ -2,22 +2,77 @@
...
+            raise ValueError('Number of channels is %d' % image.shape[-1])
+        else:
+            img_array = image
+
+        if shape is not None and img_array.shape != shape:
+            mismatch = '%s != %s' % (str(img_array.shape), str(shape))
+            raise IndexError('Image shapes do not all match: %s', mismatch)
+
+        if img_array.dtype != np.uint8:
+            if 0 <= np.min(img_array) <= 1 and 0 <= np.min(img_array) <= 1:
```
Comment (unmonoqueteclea) [inline sensenet/load.py:45]: You repeated `np.min` twice

**Review by unmonoqueteclea — CHANGES_REQUESTED** [review]
Only a little bug (I think) and a question

**Comment by unmonoqueteclea** [comment]
Does the interface change? I understand that I do not need to use, directly, the helpers functions that you included to load images because you are already taking care of that in the model wrapper.

As exported models already worked with pixel inputs, there won't be any change in that interface either.

Should I remove `input_image_format` or any other parameter from model settings or they are still being used by the wrapper?

**Comment by unmonoqueteclea** [comment]
No problem with the PIL dependency, I was already using it in many projects

**Comment by charleslparker** [comment]
> Should I remove input_image_format or any other parameter from model settings or they are still being used by the wrapper?

You can remove that and `image_file_prefix` if you were using that.  And in fact, I think it will throw an exception if you do not remove them.  If this is hard I can leave them in and they will have no effect.

**Comment by charleslparker** [comment]
If there's nothing else besides the `tensorflow-gpu` thing, could you accept this PR @unmonoqueteclea ?  I'd like to have that latter in a separate PR.

**Review by unmonoqueteclea — APPROVED** [review]
Looks great now

**Pull request merged**